### ISSUE TITLE
Optimizing particles

### ISSUE THREAD

**Issue by illwieckz** [Unvanquished/Unvanquished#2568]
Loading the default spectator view of chasm with snow flakes is a good way to test the particle code. With this scene, the CPU usage is easily higher than GPU usage.

_Note:_ The following numbers come from a recording I did today with Orbit, numbers may change given the hardware and configuration, but that gives a good idea about what's going on. Also this is _after_ #826 which already optimizes things a bit. Those numbers assume the GPU is fast, on that purpose I've set my CPU to powersave profile and my GPU to high performance. This inflates the CPU numbers but also make them easier to read. This is on dll to allow Orbit to record the game code as well, nexe code may be slower (I assume less vectorization and platform specific code, plus the virtual machine overhead).

When doing so, the CPU is most consumed in `CG_DrawActiveFrame` (43.86%), in which the one spending time is `CG_AddParticles` (36.60%).
In CG_AddParticle, the ones spending time are `CG_EvaluateParticlePhysics` (24.35%) and `CG_RenderParticle` (11.32%).
In `CG_EvaluateParticlePhysics`, the one spending time is `CG_Trace` (22.68%), itself calling more trace code recursively.
In `CG_RenderParticle`, the one spending time is `trap_R_AddRefEntityToScene` (9.47%), itself spending time in IPC.

So, any improvement to tracing would help a lot.

Any improvement in IPC would also help. I also wonder if particles can be implemented in engine or not, this would not only give higher performance code than nexen it would also avoid IPC. The particle code seems super stable in a way no one really contributes to it to change its behavior, it does the job and no one feels the need to add features. Moving it to the engine would make us lose the ability to update it by simply distributing a mod, but unless we break the API (that also look super stable), users can rebuild the engine to get updates.

**Comment by illwieckz** [comment]
Hmm, I just thought that moving particle code to engine would also require a lot of the CG code to live in engine, that may not be a good idea.

Another thing that may help the particle code is to port it (and the tracing code) to GLM, with hope the GLM functions may bring some extra optimizations we miss.

**Comment by bmorel** [comment]
don't you already did that?

**Comment by VReaperV** [comment]
> I also wonder if particles can be implemented in engine or not, this would not only give higher performance code than nexen it would also avoid IPC.

Definitely. Can be done on the GPU too, especially as we have the prerequisites of compute shader and easy storage/uniform buffer support (we could do it in vertex/fragment shaders as well, but personally I don't wanna bother with that).

**Comment by slipher** [comment]
I don't think moving particle systems to engine is a good idea. We wouldn't be able to do the tracing part there. If IPC is a problem, we can add a batch AddRefEntityToScene.

**Comment by VReaperV** [comment]
Tracing can be done in a shader against the depth buffer. There would, of course, be some edge cases where the particle should bounce off of some off-screen surface, but I don't think it would be very noticeable.
